File: backend/models/StudySession.js
import mongoose from 'mongoose';

const studySessionSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  subject: {
    type: String,
    required: [true, 'Please provide a subject'],
    trim: true
  },
  topic: {
    type: String,
    trim: true
  },
  startTime: {
    type: Date,
    required: true
  },
  endTime: {
    type: Date
  },
  duration: {
    type: Number, // in minutes
    default: 0
  },
  breaks: [{
    startTime: Date,
    endTime: Date,
    duration: Number // in minutes
  }],
  focusScore: {
    type: Number,
    min: 0,
    max: 100,
    default: 100
  },
  notes: {
    type: String,
    maxlength: [1000, 'Notes cannot be more than 1000 characters']
  },
  completed: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: true
});

// Calculate duration when session ends
studySessionSchema.pre('save', function(next) {
  if (this.endTime && this.startTime) {
    const totalMinutes = Math.floor((this.endTime - this.startTime) / (1000 * 60));
    const breakMinutes = this.breaks.reduce((sum, br) => sum + (br.duration || 0), 0);
    this.duration = totalMinutes - breakMinutes;
  }
  next();
});

const StudySession = mongoose.model('StudySession', studySessionSchema);

export default StudySession;

File: lib/models/study_session.dart
class StudySession {
  final String? id;
  final String subject;
  final String? topic;
  final DateTime startTime;
  final DateTime? endTime;
  final int duration; // in minutes
  final List<Break>? breaks;
  final int focusScore;
  final String? notes;
  final bool completed;
  final DateTime? createdAt;

  StudySession({
    this.id,
    required this.subject,
    this.topic,
    required this.startTime,
    this.endTime,
    this.duration = 0,
    this.breaks,
    this.focusScore = 100,
    this.notes,
    this.completed = false,
    this.createdAt,
  });

  factory StudySession.fromJson(Map<String, dynamic> json) {
    return StudySession(
      id: json['_id'] ?? json['id'],
      subject: json['subject'] ?? '',
      topic: json['topic'],
      startTime: DateTime.parse(json['startTime']),
      endTime: json['endTime'] != null ? DateTime.parse(json['endTime']) : null,
      duration: json['duration'] ?? 0,
      breaks: json['breaks'] != null
          ? (json['breaks'] as List).map((b) => Break.fromJson(b)).toList()
          : null,
      focusScore: json['focusScore'] ?? 100,
      notes: json['notes'],
      completed: json['completed'] ?? false,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'subject': subject,
      if (topic != null) 'topic': topic,
      'startTime': startTime.toIso8601String(),
      if (endTime != null) 'endTime': endTime!.toIso8601String(),
      'duration': duration,
      if (breaks != null) 'breaks': breaks!.map((b) => b.toJson()).toList(),
      'focusScore': focusScore,
      if (notes != null) 'notes': notes,
      'completed': completed,
    };
  }
}

class Break {
  final DateTime startTime;
  final DateTime? endTime;
  final int duration; // in minutes

  Break({
    required this.startTime,
    this.endTime,
    this.duration = 0,
  });

  factory Break.fromJson(Map<String, dynamic> json) {
    return Break(
      startTime: DateTime.parse(json['startTime']),
      endTime: json['endTime'] != null ? DateTime.parse(json['endTime']) : null,
      duration: json['duration'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'startTime': startTime.toIso8601String(),
      if (endTime != null) 'endTime': endTime!.toIso8601String(),
      'duration': duration,
    };
  }
}
